### ISSUE TITLE
no superclass method in REXML tests @ warm

### ISSUE THREAD

**Issue by mgaudet** [rubyomr-preview/ruby#59]
94462bf1dd1900a54ec4e55c24b2ed10fdf2e635
```
magaudet@ubuntu1604VM:~/open/ruby_Inliner (detached*)*$ OMR_JIT_OPTIONS=count=3,optLevel=warm LD_LIBRARY_PATH=$PWD make test-all TESTS=test/rexml
<snip>
Run options: "--ruby=./miniruby -I./lib -I. -I.ext/common  ./tool/runruby.rb --extout=.ext  -- --disable-gems" --excludes-dir=./test/excludes --name=!/memory_leak/

# Running tests:

[361/425] REXMLTests::Tester#test_element = 0.03 s                                                                                                                
  1) Error:
REXMLTests::Tester#test_element:
NoMethodError: super: no superclass method `parent=' for "1) He's a great guy!":REXML::Text
    /home/magaudet/open/ruby_Inliner/lib/rexml/text.rb:125:in `parent='
    /home/magaudet/open/ruby_Inliner/lib/rexml/parent.rb:142:in `replace_child'
    /home/magaudet/open/ruby_Inliner/lib/rexml/child.rb:30:in `replace_with'
    /home/magaudet/open/ruby_Inliner/lib/rexml/element.rb:506:in `text='
    /home/magaudet/open/ruby_Inliner/test/rexml/test_core.rb:519:in `test_element'

[365/425] REXMLTests::Tester#test_empty_doc = 0.00 s           
  2) Error:
REXMLTests::Tester#test_empty_doc:
NoMethodError: super: no superclass method `encoding_updated' for #<REXML::IOSource:0x0055a90e427340>

<snip>
```

**Comment by mgaudet** [comment]
Seeing something very similar running optcarrot: 

```
magaudet@ubuntu1604VM:~/open/optcarrot (detached*)*$ ruby -Jcount=10,optLevel=warm,verbose,limit={*clock_envelope*} bin/optcarrot  --benchmark examples/Lan_Master.nes

 compiling /home/magaudet/open/optcarrot/lib/optcarrot/apu.rb:464:clock_envelope
+ (warm @ 0x7fec54783034 t=684174ms  63.163ms) /home/magaudet/open/optcarrot/lib/optcarrot/apu.rb:464:clock_envelopefps: 0

checksum: 4096
/home/magaudet/open/optcarrot/lib/optcarrot/apu.rb:499:in `active?': super: no superclass method `active?' for #<Optcarrot::APU::Pulse> (NoMethodError)
	from /home/magaudet/open/optcarrot/lib/optcarrot/apu.rb:466:in `clock_envelope'
	from /home/magaudet/open/optcarrot/lib/optcarrot/apu.rb:175:in `clock_oscillators'
	from /home/magaudet/open/optcarrot/lib/optcarrot/apu.rb:198:in `clock_frame_counter'
	from /home/magaudet/open/optcarrot/lib/optcarrot/apu.rb:228:in `proceed'
	from /home/magaudet/open/optcarrot/lib/optcarrot/apu.rb:214:in `flush_sound'
	from /home/magaudet/open/optcarrot/lib/optcarrot/apu.rb:161:in `vsync'
	from /home/magaudet/open/optcarrot/lib/optcarrot/nes.rb:44:in `step'
	from /home/magaudet/open/optcarrot/lib/optcarrot/nes.rb:77:in `run'
	from bin/optcarrot:6:in `<main>'
```

**Comment by mgaudet** [comment]
Dramatically reduced test case here: 

```ruby
class Base
   attr_accessor :active

   def active?
      return true
   end

   def step
      @active = active?
   end

end


class ImplA < Base
   def active?
      super
   end
end


class ImplB < Base
   def active?
      super
   end
end

a = ImplA.new
b = ImplB.new

4.times { 
   a.step
   b.step
}
```

When compiling step, the inliner looks in the call cache for `active?`, and inlines the last called version. Then the next time around, the guard fails, and we call into the other impl. Then we come back around, and the guard passes... and here the call to super fails. 

```
magaudet@ubuntu1604VM:~/open/ruby_Inliner (cleanup)$ LD_LIBRARY_PATH=$PWD ./ruby -Jcount=1,optLevel=warm --disable-gems ../matt_tests/inlining_call_to_super.rb
../matt_tests/inlining_call_to_super.rb:24:in `active?': self has wrong type to call super in this context: ImplB (expected ImplA) (TypeError)
	from ../matt_tests/inlining_call_to_super.rb:9:in `step'
	from ../matt_tests/inlining_call_to_super.rb:33:in `block in <main>'
	from ../matt_tests/inlining_call_to_super.rb:31:in `times'
	from ../matt_tests/inlining_call_to_super.rb:31:in `<main>'
```

It's not clear why to me yet though.

**Comment by mgaudet** [comment]
Resolved by #63
